[advanced digitizing] Add circles intersection digitizing tool

nirvn opened the pull request:
## Description

This PR adds a brand new advanced digitizing tool allowing for users to pick/digitize a point at the intersection of two circles. In action:

https://github.com/qgis/QGIS/assets/1728657/fdef5525-9777-4bb7-92dd-ad24f14a2701

The implementation adds an abstract advanced digitizing tool class to allow for more such "COGO" tools to be added in the future (e.g. intersection between a circle and a line + bearing).

The new functionality is exposed to our python bindings, which also means people can write plugins and prototype additional functionalities in python.

(The PR temporarily includes the construction guides work, I'll rebase when that is merged)

nirvn committed:
[advanced digitizing] Add circles intersection digitizing tool

nirvn committed:
Add missing documentation

uclaros commented:
While I agree that this functionality is missed in QGIS, the UX of this implementation is not really improved over the previous attempt https://github.com/qgis/QGIS/pull/45904

Currently, a single circle constraint exists by using the _Distance_ constraint, and imho we should take advantage of that and use it at least for one of the two intersecting circles.

How about we add the distance circles as _guides_ too, then allow snapping on their intersections?

lbartoletti commented:
I'm +1 to have this kind of tool in QGIS, but we were refused this one because we were adding a CoGO toolbar (later, I think there were other toolbars, but hey ...). 

We finally made a plugin for it, with other CoGo tools that I didn't release because we moved on to other subjects and the spirit of the project at the time no longer suited me or my colleague... Tired of the feeling of differences in treatment, we didn't open the QEP, since we felt that certain more important changes were often made without this "precious discussion"...


Nevertheless, I'll support you @nirvn so that we can finally have this kind of tool in QGIS! And I'm willing to come back and support you. Let's see how we can mutualize this with a UI/UX that suits us and our use cases.

nirvn commented:
Good reminder bot, I'll push  this forward over the weekend.

nirvn commented:
@uclaros , hey, sorry for taking so long to engage with your comment, life and priorities got in the way :)

While I appreciate what you are suggesting here, I do not think a COGO tool should _rely_ on construction guides to work. Very often, these operations are one off. Most basic case here: I'm back from my field survey where I took accurate reading of a pair of reachable points and distances to map out the intersecting point. I've got no need here for construction guides as these two circles will never be re-used. 

So IMHO, I would argue here that COGO operations, very often ephemeral, should be kept separate from construction guides. 

What I very much _do_ like in your suggestion is adding the possibility to add circular construction guides when the distance is looked during construction mode. That's something we should consider in a separate PR.

nirvn commented:
@uclaros , @lbartoletti , I do feel this is a UI improvement from #45904 in quite a few fronts here:

1/ the big difference here is that it integrates *with* the CAD tool and ends up adding _a vertex_ to any map tool that shares the CAD tool class. This means support for adding a vertex using this tool while digitizing line or a polygon (think about the need to trace a plot of land that includes circle-intersection points). The previous PR was a separate process that did not plug into the existing digitizing framework, only worked for point geometry layers, and didn't play along other map tools (e.g. annotations).

2/ the UI does not involve adding a new toolbar in our main windows but rather add to the CAD panel, which IMHO gives it a better location while avoiding cluttering the main window toolbar area.

@lbartoletti , sorry you were left frustrated by the process and glad to read I've got your support on this one. On the plugin side of things, the PR allows for other tools to be prototyped and registered via plugin. If your unreleased COGO plugin included other tools, I'd love to see if it's possible to port them to this framework. It'd make for a great QGIS 3.40 plugin :)

nirvn commented:
A screencast of what I was explaining above in terms of integration with map tools:

https://github.com/user-attachments/assets/1b51f337-92db-43d0-8f5a-97664e58b6d9

Here, I was not able to get the top-right corner of the rice field (too muddy), so I had to take two points + distances. The video demonstrates how I could nevertheless go back to the office and create the plot  polygon by using the COGO tool to add the 4th vertex.

nirvn committed:
Add advanced digitizing tools registry

nirvn commented:
@nyalldawson , I've (partially) addressed the comment you made in #45904 by adding a registry of advanced digitizing tools (that'll be useful to you @lbartoletti ) and transforming the toolbar action into a drop down menu that will contain _all_ the registered tools.

ATM, I think I'm -1 on doing the gimp-like subpanel because a/ we have only one tool, and more importantly b/ these tools can still rely on the CAD constraint widgets, so opening a subpanel would just interfered with that.

I do think the toolbar drop down menu does embrace the spirit at least of a tools container :)

nyalldawson commented on src/gui/qgsadvanceddigitizingcanvasitem.cpp:
```diff
@@ -59,6 +59,13 @@ void QgsAdvancedDigitizingCanvasItem::paint( QPainter *painter )
   }
 
+  // Draw current intersection tool
+  if ( mAdvancedDigitizingDockWidget->tool() )
```
```suggestion
  if ( QgsAdvancedDigitizingTool * tool = mAdvancedDigitizingDockWidget->tool() )
  {
      // if a tool is active in the dock, then delegate to that tool to handle decorating the canvas instead of using the default decorations
      tool->paint( painter );
      return;
  }
```

(Maybe in future we could add some flag for the tool "draw default CAD decorations", when there's a need to)

nyalldawson commented on src/gui/qgsadvanceddigitizingdockwidget.cpp:
```diff
@@ -237,6 +240,22 @@ QgsAdvancedDigitizingDockWidget::QgsAdvancedDigitizingDockWidget( QgsMapCanvas *
 
+  // Tools
+  QMenu *toolsMenu = new QMenu( this );
+  const QStringList toolMetadataNames = QgsGui::instance()->advancedDigitizingToolsRegistry()->toolMetadataNames();
```
If this is built on QgsAdvancedDigitizingDockWidget construction then it'll never react to plugin installs/changes. I'd move this logic to a slot connected to toolsMenu::aboutToShow, and populate the menu at time of display instead.

nirvn replied:
@nyalldawson , good idea; I'm also adding signals to the register but here the about to show is a much better way to go.

nyalldawson commented on src/gui/qgsadvanceddigitizingdockwidget.cpp:
```diff
@@ -644,6 +676,29 @@ void QgsAdvancedDigitizingDockWidget::activateCad( bool enabled )
+    mCurrentTool = nullptr;
+  }
+
+  qDebug() << mCurrentTool;
```
```suggestion
```

nyalldawson commented on src/gui/qgsadvanceddigitizingdockwidget.cpp:
```diff
@@ -1492,6 +1547,70 @@ QList<QgsPointXY> QgsAdvancedDigitizingDockWidget::snapSegmentToAllLayers( const
+{
+  if ( mCurrentTool )
+  {
+    mCurrentTool->canvasPressEvent( event );
```
I think this should check:

   if ( event->isAccepted() )
      return;

elsewhere too, so that the subclasses can determine whether events should be propogated or not...

nyalldawson commented on src/gui/qgsadvanceddigitizingdockwidget.h:
```diff
@@ -271,6 +276,36 @@ class GUI_EXPORT QgsAdvancedDigitizingDockWidget : public QgsDockWidget, private
 
+    /**
+     * Processes the canvas press \a event.
+     * \returns Returns TRUE if the event has been 'eaten' and should not propagate further.
```
As noted before, it'd be cleaner to use event->isAccepted()/isIgnored() here to determine propagation.

nyalldawson commented on src/gui/qgsadvanceddigitizingtools.h:
```diff
@@ -0,0 +1,158 @@
+ * \brief An abstract class for advanced digitizing tools.
+ * \since QGIS 3.40
+ */
+class GUI_EXPORT QgsAdvancedDigitizingTool : public QWidget
```
I don't think this should be a QWidget -- I highly suspect that choice would bite us in future. Rather there should be a member which creates the widget when required instead.

```suggestion
class GUI_EXPORT QgsAdvancedDigitizingTool : public QObject
```

nyalldawson commented on src/gui/qgsadvanceddigitizingtools.h:
```diff
@@ -0,0 +1,158 @@
+
+    /**
+     * Handles canvas press event. If TRUE is returned, the tool will have
+     * blocked the event for propagating.
```
Again, i'd use isAccepted/isIgnored for propagation

nyalldawson commented on src/gui/qgsadvanceddigitizingtoolsregistry.h:
```diff
@@ -0,0 +1,153 @@
+/**
+ * \ingroup gui
+ * \brief Registry of available advanced digitizing tools.
+ *
```
Some note about being accessed via QgsGui::advancedDigitizingToolsRegistry()
{

nyalldawson commented:
@nirvn Ok, i've done a review based on the code as is. I'm fine with the new GUI approach here too, we can always revisit the tool box concept in future if/when needed.

Keen to hear @uclaros / @lbartoletti 's thoughts on the new approach too!

nirvn committed:
Address review and add test

nyalldawson commented on src/gui/qgsadvanceddigitizingdockwidget.cpp:
```diff
@@ -237,6 +240,25 @@ QgsAdvancedDigitizingDockWidget::QgsAdvancedDigitizingDockWidget( QgsMapCanvas *
+    for ( const QString &name : toolMetadataNames )
+    {
+      QgsAdvancedDigitizingToolAbstractMetadata *toolMetadata = QgsGui::instance()->advancedDigitizingToolsRegistry()->toolMetadata( name );
+      QAction *toolAction = new QAction( toolMetadata->icon(), toolMetadata->visibleName(), this );
```
```suggestion
      QAction *toolAction = new QAction( toolMetadata->icon(), toolMetadata->visibleName(), toolsMenu );
```

nyalldawson commented on src/gui/qgsadvanceddigitizingdockwidget.cpp:
```diff
@@ -237,6 +240,25 @@ QgsAdvancedDigitizingDockWidget::QgsAdvancedDigitizingDockWidget( QgsMapCanvas *
+  // Tools
+  QMenu *toolsMenu = new QMenu( this );
+  connect( toolsMenu, &QMenu::aboutToShow, this, [ = ]()
+  {
```
```suggestion
  {
     toolsMenu->clear();
```

uclaros commented:
@nirvn Thanks for being open to discussion!
I agree with all your points above and the UI is improved (cad tools integration was a must) however the UX is similar.: the user still needs to do lots of clicks to get what he wants. Your screencasts display a slightly awkward workflow: _point, point, distance, distance_ while I think the UI should be optimized for: _point, distance, point, distance_.

I'd be very happy with the resulting UX if the following were tuned:

1. `QgsAdvancedDigitizingCirclesIntersectionTool::createWidget()` should start with `mCircle1Digitize->setChecked( true );` so that the user can directly pick his first point once the tool is activated.
2. Once a point is picked, `mCircle1Distance` / `mCircle2Distance` should gain focus respectively
3. Once `mCircle1Distance` is accepted (enter pressed), `mCircle2Digitize` should be triggered
4. At the beginning of `QgsAdvancedDigitizingCirclesIntersectionTool::createWidget()`, we could check `mCadDockWidget->constraintDistance()` and if one is set, pre-populate `mCircle1...` widgets, and continue with having `mCircle2Digitize->setChecked( true );` so the user can directly pick his second point
5. If the project CRS is not projected, the tool should probably be disabled
6. It would be great if we could have a bindable shortcut for activating the tool.

lbartoletti commented:
@nirvn Thanks.

I'm not able to review and test this PR at the moment. But, be sure that, I'll be happy to review and test at the end of August.

nirvn committed:
Apply requested UX logic

nirvn commented:
@uclaros , thanks for the feedback. I've implemented the UX logic you've described (point 1 through to 4). 

With regards to project CRS, I'll think of a way to do that in a follow up PR (I'm thinking of a flags

uclaros commented:
Thanks @nirvn for handling my suggestion. However I do not see how 4 is implemented, am I missing something?
This is important imho as it allows for first circle to be placed at the last digitized segment, which is probably the most common use case.

Regarding 3, we should probably first check if point 1 has been set, or the user might be trying to do _distance, point, distance, point_

I didn't get a chance to build and test yet, but shouldn't `mCircle1Digitize` and `mCircle2Digitize` buttons be mutually exclusive?

nirvn committed:
[cad] Make 2-circle intersection tool's digitizing buttons mutually exclusive

nirvn committed:
[cad] Implement locked distance logic for 2-circle intersection tool

nirvn commented:
@uclaros , heh, I had meant 1-3, but then you insisting on 4 led to it being implemented ;)

Also, I've made the digitize toggles mutually exclusive, good call. As for the rest, let's tweak things in follow up PRs. I'm not sure we should overthink the logic too much here though.

That said, big thanks for the feedback, it made the feature much better :pray:

uclaros commented on src/gui/qgsadvanceddigitizingtools.cpp:
```diff
@@ -0,0 +1,299 @@
+  connect( mCircle1Y, &QgsDoubleSpinBox::textEdited, this, [ = ]() { mCircle1Digitize->setChecked( false ); } );
+  layout->addWidget( mCircle1Y, 2, 1 );
+
+  label = new QLabel( QStringLiteral( "Distance" ), toolWidget );
```
`Distance` is too long string (also not translated) causing the ugly widget to grow unnecessarily, let's change that to `R` for radius or `d` for distance, which is what is used in the advanced digitizing panel.

uclaros commented on src/gui/qgsadvanceddigitizingtools.cpp:
```diff
@@ -0,0 +1,299 @@
+  connect( mCircle2Y, &QgsDoubleSpinBox::textEdited, this, [ = ]() { mCircle2Digitize->setChecked( false ); } );
+  layout->addWidget( mCircle2Y, 6, 1 );
+
+  label = new QLabel( QStringLiteral( "Distance" ), toolWidget );
```
same

uclaros reviewed (commented):
@nirvn I would have appreciated it if you waited for the review process to complete before merging this!

A couple other issues that need to be resolved before I'd stop considering this half-baked:
- Handle `esc` button: users are trapped in this mode and cannot abort unless they do add a point.
- Handle right click: right click should do the same as `esc`, right now it forwards the right click to the map tool
- Once a center is placed and focus goes into its radius widget, mouse move should track the distance from that center and update the spinbox value, similar to how `d` works for advanced digitizing panel.

nirvn commented:
@uclaros , good catch regarding the ESC and right-click handling, I've addressed these alongside the label renaming suggestions you dropped in this follow up PR: https://github.com/qgis/QGIS/pull/58289

As for merging this to begin with, after having answered to reviews and implemented UI/UX suggestions from you and others over the last 7 weeks, without an explicit request for further review I think it was fair to merge to get as much testing as we can. On a friendly note,  I'm not sure labeling efforts shown this PR as "half-baked" is a constructive way to go about it nor does it reflect the willingness to adopt a fair amount of what was discussed here ;)

On defining the map canvas overlay widget as "ugly", what is used here is what most other map tools do (i.e. e.g. offset curve map tool, rotation map tool, scale map tool, etc.). If you feel like there's room for improvements for this well-established overlay, this can be discussed separately. We could give some tweaks a try (e.g. semi-transparent background?), but we're still a bit limited by what Qt widgets offers us. 

Thanks for the reviews here, you've made this a better feature; keep them coming :pray:

agiudiceandrea commented:
Hi @nirvn, please see the issue report at https://github.com/qgis/QGIS/issues/59039.

DelazJ commented:
@nirvn It looks weird (to me) to have a drop-down menu button to access ... a single button. I don't remember any place we did this. Could we not add the drop-down menu only when there are more children? It would ease access to the tool in the time-being (one less click).
Moreover, I'd like to make a call I did some years ago; please, guys, name or add tooltips to the tools, it helps a lot when you need to document and name them. So, please, could the drop-down menu there get a tooltip? Thanks.
